20231214-listUsers-api.md

```diff
@@ -0,0 +1,887 @@
+  schema 1.1
+
+type user
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [group#member]
+```
+| object     | relation | user             |
+|------------|----------|------------------|
+| document:1 | viewer   | group:eng#member |
+| group:eng  | member   | group:fga#member |
+```
+ListUsers({
+  object: "document:1",
+  relation: "viewer",
+  user_filter: [
+    {
+      type: "group",
+      relation: "member"
+    }
+  ]
+}) --> {
+  users: ["group:eng", "group:fga"],
+  wildcard_types: []
+}
+```
+This example deviates from many of the examples above in that we expand all relationships for a specific object and relation (e.g. `document:1#viewer`) that are related to a given set of users or subject set (e.g. `group#member`).
+
+1. Expand the set of subjects/users that relate to `document:1#viewer`.
+
+   `expand(document:1#viewer)` --> ["group:eng#member"]
+
+   1. We find `group:eng#member`, which matches the `group#member` in the user filter, so add `group:eng` to the list of items in the response.
+
+1. Even though we found `group:eng#member` and it matched the `group#member` user filter, we continue expanding the set of subjects/users that relate to `group:eng#member` because `group#member` is related to `group#member` and thus there may be more nested groups matching the specific filter.
+
+   `expand(group:eng#member)` --> ["group:fga#member"]
+
+    1. We find `group:fga#member`, which matches the `group#member` in the user filter, so add `group:fga` to the list of items in the response.
+
+1. Again, we continue expansion because we may find more specific nested groups that relate to `group:fga#member`.
+
+   `expand(group:fga#member)` --> []
+
+1. No further subjects to expand, and so we're done. Return the items we accumulated from the steps above.
+
+Visually, the overall recursive call tree looks like the following:
+```
+expand(document:1#viewer) --> ["group:eng#member"]
+yield group:eng
+|-> expand(group:eng#member) --> ["group:fga#member"]
+    yield group:fga
+|---> expand(group:fga#member) --> []
+
+return ["group:eng", "group:fga"]
+```
+
+### Intersection and Exclusion
+For relationships that involve an intersection (e.g. `a and b`) or exclusion (e.g. `a but not b`) we'll apply the same algorithmic approach we do in ListObjects. Namely, given the set `a and b` we'll compute `a` and then call `Check` for each of the results to resolve the set intersection through Check resolution. Likewise, for `a but not b` we'll compute `a` and then call `Check` for each of the results to resolve the set difference.
+
+This choice is an algorithmic choice that exploits the fact that `a and b` and `a but not b` can be no larger than the `max(size(a), size(b))`, so instead of computing the results for both sets, holding the results temporarily in memory and then resolving the overlap, we simply compute the first set and then use Check resolution to resolve the residual.
```

Comment:
Can we have some performance measurements on this assumption?

Reply:
I think that's out of scope for now. It was out of scope for ListObjects too 😅